benchflow-test-manager/application/src/main/java/cloud/benchflow/testmanager/strategy/selection/SelectionStrategy.java

```diff
@@ -1,17 +1,100 @@
+
 /**
  * @author Jesper Findahl (jesper.findahl@usi.ch) created on 2017-04-20
  */
-public interface SelectionStrategy {
+public abstract class SelectionStrategy {
+
+  private final MinioService minioService;
+  private final ExplorationModelDAO explorationModelDAO;
+  private Logger logger = LoggerFactory.getLogger(SelectionStrategy.class.getSimpleName());
+
+  public SelectionStrategy(Logger logger) {
+
+    this.logger = logger;
+
+    this.minioService = BenchFlowTestManagerApplication.getMinioService();
+    this.explorationModelDAO = BenchFlowTestManagerApplication.getExplorationModelDAO();
+  }
+
+  // only used for testing
+  public SelectionStrategy(Logger logger, MinioService minioService,
+      ExplorationModelDAO explorationModelDAO) {
+    this.logger = logger;
+    this.minioService = minioService;
+    this.explorationModelDAO = explorationModelDAO;
+  }
+
+  public SelectedExperimentBundle selectNextExperiment(String testID) {
+
+    logger.info("selectNextExperiment: " + testID);
+
+    try {
+
+      // get executed exploration points
+      List<Integer> explorationPointIndices =
+          explorationModelDAO.getExecutedExplorationPointIndices(testID);
+
+      String testDefinitionYamlString =
+          IOUtils.toString(minioService.getTestDefinition(testID), StandardCharsets.UTF_8);
+
+      String deploymentDescriptorYamlString = IOUtils
+          .toString(minioService.getTestDeploymentDescriptor(testID), StandardCharsets.UTF_8);
+
+      // JavaCompatExplorationSpace explorationSpace = explorationModelDAO.getExplorationSpace(testID);
+      ExplorationSpaceGenerator.ExplorationSpace explorationSpace =
+          ExplorationSpace.explorationSpaceFromTestYaml(testDefinitionYamlString);
+
+      // next experiment to be executed
+      int nextExplorationPoint =
+          getNextExplorationPoint(explorationPointIndices, explorationSpace.size());
+
+      // get the experiment bundle for the given point
+      // <ExperimentDefinition, DeploymentDescriptor>
+      Tuple2<String, String> experimentBundle =
+          ExplorationSpace.generateExperimentBundle(explorationSpace, nextExplorationPoint,
+              testDefinitionYamlString, deploymentDescriptorYamlString);
+
+      return new SelectedExperimentBundle(experimentBundle._1(), // experiment definition
+          experimentBundle._2(), // deployment descriptor
+          nextExplorationPoint);
+
+
+    } catch (BenchFlowTestIDDoesNotExistException | BenchFlowDeserializationException
+        | IOException e) {
+      // should not happen
+      // TODO - handle me
+      e.printStackTrace();
+    }
+
+    return null;
+
+  }
+
+  protected abstract int getNextExplorationPoint(List<Integer> executedExplorationPointIndices,
+      int explorationSpaceSize);
 
-  enum Type {
+  public enum Type {
     ONE_AT_A_TIME, RANDOM_BREAKDOWN, BOUNDARY_FIRST
   }
 
-  SelectedExperimentBundle selectNextExperiment(String testID);
+  public class SelectedExperimentBundle {
```

Comment:
It is used by other classes outside the package. We can also separate it into its own class if you think that is better. It is now there just because it is related to the `SelectionStrategy`